### PR TITLE
feat(url_helper): add default timeout to readurl() to prevent hanging

### PR DESCRIPTION
**Pull request by a-dubs**
## Proposed Commit Message

```
feat(url_helper): add default timeout to readurl() to prevent hanging

If there is an issue with networking, we don't want any calls to hang 
forever. This is never a desireable outcome, because this prevents 
cloud-init from finishing and thus makes debugging much more difficult.
```

## Additional Context
This issue was discovered when attempting to contact the IMDS from an
oracle VM that had failed setting up its networking.  

## Checklist

- [X] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/development/index.html)
- [X] I have updated or added any [unit tests](https://cloudinit.readthedocs.io/en/latest/development/testing.html) accordingly
- [X] I have updated or added any [documentation](https://cloudinit.readthedocs.io/en/latest/development/contribute_docs.html) accordingly

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

**Review by TheRealFalcon — COMMENTED** [review]
LGTM with an inline comment.

Let's get a +1 from one of the other committers that hasn't been staring at this code all week :zany_face:

File: cloudinit/url_helper.py
Comment (TheRealFalcon) [inline cloudinit/url_helper.py:196]: I found today that the [EC2 datasource](https://github.com/canonical/cloud-init/blob/main/cloudinit/sources/DataSourceEc2.py#L90) sets a default timeout to 50, so I think we should bump this up to 60 as a global default.

Comment (a-dubs) [reply cloudinit/url_helper.py:196]: 50s? Are their server rooms full of carrier pigeons??? 
Just kidding. I agree 60s is good!

**Review by holmanb — CHANGES_REQUESTED** [review]
> we don't want any calls to hang forever. This is never a desireable outcome

Yeah this would probably have made debugging that issue easier. I agree that this proposal is probably fine for most cloud environments, but I'm a little worried about regression potential of this kind of change. How confident are we that "never" really means "never"?

I'm imagining a hypothetical workload which might kick off a cloud-init instance to start, simultaneously fire off building an imds from scratch (or something that might take really long to finish), and let the blocked instance wait "indefinitely" (30 minutes, 2 hours, idk) until the server is available and serving before the instance would continue.

Are we sure that this can't happen in any cloud-init codepaths? I see that NoCloud uses a timeout of 5s in `util.read_seeded()`, so that shouldn't be a culprit, but are there any other code paths that might do something like this?

Some scientific computing clusters take a really long time to boot (think minutes to hours), and I know that cloud-init is now used in many of those environments.

I'd like to make sure that we're really confident that we **never** want to wait forever before we change the default to a timeout. 
I'm -1 to this changeset for now, but happy to be convinced otherwise - I'm probably being too cautious.

**Comment by a-dubs** [comment]
@holmanb Yeah good point. I should really _never_ use "never" like that. 

But yeah I'm honestly okay with keeping it cloud-specific for now. For oracle, I'm quite certain a small timeout is reasonable so we'll just put the timeout there - I believe @TheRealFalcon is working on a PR that will add short timeout (5s) to the oracle datasource's calls to the IMDS via read_url().

Also, I am not sure I understand how this timeout would matter in terms of boot speed of the instance it's running on though? Since this is called against the datasource, it doesn't matter how long the instance would take to boot because once it boots the datasource should respond in a normally timely manner, right? Or I am misunderstanding? Thanks!

**Comment by holmanb** [comment]
> Also, I am not sure I understand how this timeout would matter in terms of boot speed of the instance it's running on though? Since this is called against the datasource, it doesn't matter how long the instance would take to boot because once it boots the datasource should respond in a normally timely manner, right? 

I wouldn't expect the timeout to negatively affect boot speed.

> Or I am misunderstanding? Thanks!

I wasn't trying to raise concerns about boot speed. My concerns were actually about the reliability (and changing behavior) for systems that boot slowly and therefore depend on cloud-init not having a timeout.


Question:

@a-dubs When systems don't use ds-identify, in theory they have to iterate through datasources until they find one that responds. Wouldn't oracle's "no timeout" have broken this design? Or am I missing something? In theory this doesn't really matter too much anymore, since most (all?) supported init systems now have support for ds-identify, but I'm curious if this would have broken that model.

**Comment by a-dubs** [comment]
> Question:
@a-dubs When systems don't use ds-identify, in theory they have to iterate through datasources until they find one that responds. Wouldn't oracle's "no timeout" have broken this design? Or am I missing something? In theory this doesn't really matter too much anymore, since most (all?) supported init systems now have support for ds-identify, but I'm curious if this would have broken that model.

That's a good question. For Oracle specifically, the  readurl() call against the imds that has no timeout currently, is done once Oracle is determined to be the datasource. Let's sync up on this @holmanb because it sounds like we both need to just clarify a few things. I know I'd benefit from it at least.

**Comment by TheRealFalcon** [comment]
Good points @holmanb . +1 to closing this PR.

**Comment by holmanb** [comment]
>> Wouldn't oracle's "no timeout" have broken this design? Or am I missing something?
>
> DataSourceOracle has a custom ds_detect() which eventually checks the [chassis-asset-tag](https://github.com/canonical/cloud-init/blob/main/cloudinit/sources/DataSourceOracle.py#L341) . If this returns False, _get_data() isn't called.

I should have checked the code before asking. That explains it.

**Comment by TheRealFalcon** [comment]
Closing for the reasons mentioned by holmanb . If there's something left to be done here, feel free to re-open.

**Pull request closed without merging**
